runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/QueryablePipeline.java

```diff
@@ -123,15 +125,24 @@ private QueryablePipeline(Collection<String> transformIds, Components components
   @VisibleForTesting
   static Collection<String> getPrimitiveTransformIds(RunnerApi.Components components) {
     Collection<String> ids = new LinkedHashSet<>();
+
     for (Map.Entry<String, PTransform> transformEntry : components.getTransformsMap().entrySet()) {
       PTransform transform = transformEntry.getValue();
       boolean isPrimitive = isPrimitiveTransform(transform);
       if (isPrimitive) {
-        List<String> subtransforms = transform.getSubtransformsList();
-        if (subtransforms.isEmpty()) {
-          ids.add(transformEntry.getKey());
-        } else {
-          ids.addAll(subtransforms);
+        // for a "primitive" transform, we want to keep only "leaf" sub-transform descendents, if any (in the common
```

Comment:
Why would a primitive transform ever have leafs?

Reply:
[This write-up on BEAM-5441](https://issues.apache.org/jira/browse/BEAM-5441?focusedCommentId=16622983&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16622983) gives the best context:

> DirectRunner [replaces a GBK](https://github.com/apache/beam/blob/master/runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectRunner.java#L264-L266) (normally a primitive transform) with [a wrapper for two runner-specific transforms ("group by key only" and "group also by window")](https://github.com/apache/beam/blob/master/runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/DirectGroupByKey.java#L63-L70), and the expectation is that the wrapped transforms are considered "primitives", but the wrapper is not
>
> … in the wordcount case, the `Reshuffle` gets added as a primitive transform (being a child of the otherwise-primitive `Read`) and, separately, its child `RemoveRandomKeys` (URN `urn:beam:transform:pardo:v1`) gets marked primitive, so there are two "primitive" transforms that have the same output PCollection, failing [the assertion](https://github.com/apache/beam/blob/9d0b03ad736797c33fbd6a4eb13155fa91367779/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/QueryablePipeline.java#L176-L182).

Basically I've been collecting cases where "primitive" transforms end up with descendents, and trying to do the right thing here as each comes to light!

Definitely interested in guidance from folks like yourself that have a more complete "big picture" understanding of whether "primitives" should never have children (let alone deeper descendents; this was the original definition of "primitive" pre-#6328, but seemed to have false-positives and -negatives), or how to handle the cases where they do (if not how I've prototyped here).

Reply:
Thanks for the pointer. so a runner might do substitution and implement a primitive in terms of other primitives, but leave the original in the graph... 

My take is that the graph should be correct (by some stricter definition) when it's submitted, but a runner can do things that make it "less" correct and we should perhaps not have as strict of checks after the fact. Treating transforms with primitive URNs but sub-transforms as composites makes sense here. 

FWIW, it's worth getting this in to unblock things.

Reply:
@ryan-williams it may be good to refer to that JIRA in the comment for context.

Reply:
@robertwb

> _a runner_ might do substitution and implement a primitive in terms of other primitives, _but leave the original in the graph_...

(mostly to check my own understanding): here it's the python SDK, not a runner, that:
1. implements `Read` (normally a primitive) in terms of other transforms (including a `Reshuffle`, which in turn wraps a `ParDo` primitive),
2. leaves the original in the graph

The _runner_ substitution case (`ReferenceRunner`) does not leave the original in the graph, and the error we were getting there had to do with a PCollection being left with no "primitive" transform that produced it.

@tweise good idea, I've updated the comment and force-pushed it here